### ISSUE TITLE
jose-jwk: `legacy::JwkEcKey` is too strict in what it accepts

### ISSUE THREAD

**Issue by Erik1000** [RustCrypto/JOSE#86]
As per [RFC7517 section 4](https://datatracker.ietf.org/doc/html/rfc7517#section-4) (fourth paragraph)
> Additional members can be present in the JWK; if not understood by
   implementations encountering them, they MUST be ignored.

But currently, if you try to parse this key for example (using the `p256` crate; key created by [step-cli](https://github.com/smallstep/cli)):
```json
{
  "use": "enc",
  "kty": "EC",
  "kid": "UYa89vgc4u_lpcbbmDQfYJQRAUD4AED8H8FUNjk5KyQ",
  "crv": "P-256",
  "alg": "ECDH-ES",
  "x": "hFc6OfbgRsYFOWyhGbWH0sS5DZBjJLGABJvPttVZfA4",
  "y": "tnXB8ks0-AZJKOgbMWJrE5Jm3nTFy0UiqQugmx9jku4"
}
```

It will fail, because it does not recognize the `use` parameter (it doesn't recognize some other parameters either, but that is the first parameter it sees).
In fact, you have to reduce it to the following parameters (so only the ones from [`Field`](https://docs.rs/elliptic-curve/latest/src/elliptic_curve/jwk.rs.html#370-376) are present)

```json
{
  "kty": "EC",
  "crv": "P-256",
  "x": "hFc6OfbgRsYFOWyhGbWH0sS5DZBjJLGABJvPttVZfA4",
  "y": "tnXB8ks0-AZJKOgbMWJrE5Jm3nTFy0UiqQugmx9jku4"
}
```

There's a registry for all key parameters at [IANA](https://www.iana.org/assignments/jose/jose.xhtml#web-key-parameters) but note that this list  1) may grow with new entries and 2) implementations are allowed to define custom parameters not in the registry.

My proposed solution would be to ignore unknown parameters just like the spec dictates it.

**Comment by ubamrein** [comment]
I was wondering what the state here is. 

At the very least the `use` field should be supported. Though it is never mentioned directly, many examples on popular RFCs contain the `use` parameter (c.f. discussion https://github.com/openid/OpenID4VCI/pull/618) in their examples.

Best would probably be to support something like https://serde.rs/attr-flatten.html#capture-additional-fields to allow deserializing and access to any additional attribute on the JWK.

**Comment by Erik1000** [comment]
I think the whole parsing from Json Web Keys and the abstraction over different EC key types does not belong into this crate, tbh.
You can have a look how we implemented JsonWebKeys in `jose` https://docs.rs/jose/latest/jose/struct.JsonWebKey.html

**Comment by tarcieri** [comment]
We could consider removing them, now that there are many other options

**Comment by Erik1000** [comment]
Yes, perhaps instead `jose` should provide a way to get the inner JwkEcKey if the rustcrypto backend is being used.

**Comment by ubamrein** [comment]
Yeah I was wondering, how one then would use e.g. the `p256` crate. Because until now I loved the ease of access to get a verification key from a JWK.

**Comment by Erik1000** [comment]
@ubamrein what is your use case? I opened the issue because we originally did not want to manually parse the x, y and d parameters (but we now do anyway...).
https://github.com/minkan-chat/jose/blob/cbde85b6a2780e96728ad1c1cecbe9bade64cbd5/src/crypto/ec.rs#L375-L382
via
https://github.com/minkan-chat/jose/blob/cbde85b6a2780e96728ad1c1cecbe9bade64cbd5/src/crypto/backend/rust/ec.rs#L99-L120
edit: sorry github did not load your response

**Comment by ubamrein** [comment]
I agreee that processing of the whole JWK RFC should be out of scope. But constructing a Public/Private key from a JWK is something I really appreciate when working with the rust-crypto crates. 

So maybe the parsing could just be made more lenient? Like ignore anything but `x`, `y`, `d` , `kty` and `crv` or so?

**Comment by Erik1000** [comment]
I think the RFC specifies parameters that MUST BE understood and I agree that other parameters (OPTIONAL ones) should be simply ignored if not understood. 
Thats what we currently do as well in the crypto primtive types in `jose`.
You should then be able to use the `JwkEcKey` together with `#[serde(flatten)]` to create your own struct with the `use` parameter:

```rust
#[derive(Deserialiaze)]
struct MyJwkEc {
     #[serde(flatten)]
     key: JwkEcKey,
     use: KeyUsage,
}

// from jose crate
#[non_exhaustive]
#[derive(Debug, Clone, Hash, PartialEq, Eq)]
pub enum KeyUsage {
    /// The `sig` (signature) value
    Signing,
    /// The `enc` (encryption) value
    Encryption,
    /// Some other case-sensitive [`String`] that did not match any of the
    /// publicly known variants
    Other(String),
}

impl Serialize for KeyUsage {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        match self {
            Self::Signing => "sig",
            Self::Encryption => "enc",
            Self::Other(s) => s,
        }
        .serialize(serializer)
    }
}

impl<'de> Deserialize<'de> for KeyUsage {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let val = <alloc::borrow::Cow<'_, str>>::deserialize(deserializer)?;
        Ok(match &*val {
            "sig" => Self::Signing,
            "enc" => Self::Encryption,
            _ => Self::Other(val.into_owned()),
        })
    }
}
```

**Comment by ubamrein** [comment]
Sure, but thats 30 lines of code vs. 1 line of code ;)

**Comment by Erik1000** [comment]
I dunno what you mean, all thats needed is this: 
```rust
#[derive(Deserialiaze)]
struct MyJwkEc {
     #[serde(flatten)]
     key: JwkEcKey,
     use: String,
}
```
if you do not want a typed use parameter (but we do!)

Of course, this will only work if JwkEcKey ignores unknown fields in its implementation

**Comment by Erik1000** [comment]
@ubamrein you could open a PR to relax deserialization of JwkEcKey and therefore close this issue

**Comment by tarcieri** [comment]
The current implementation avoids `serde_derive` but I think it's fine to switch to the custom derive which would ignore unknown fields automatically

**Comment by ubamrein** [comment]
> [@ubamrein](https://github.com/ubamrein) you could open a PR to relax deserialization of JwkEcKey and therefore close this issue

There is already a PR that's why I asked what the state is

**Comment by tarcieri** [comment]
Aah yes RustCrypto/traits#1547, which stalled with me asking if we should retire the implementation entirely 😉

**Comment by ubamrein** [comment]
> Aah yes [#1547](https://github.com/RustCrypto/traits/pull/1547), which stalled with me asking if we should retire the implementation entirely 😉

exactly :D

**Comment by tarcieri** [comment]
An updated PR like that which switches to custom derive would be fine

**Comment by tarcieri** [comment]
There's also https://github.com/RustCrypto/JOSE/tree/master/jose-jwk although offhand I'm not sure if it provides equivalent functionality

**Comment by Erik1000** [comment]
> There's also https://github.com/RustCrypto/JOSE/tree/master/jose-jwk although offhand I'm not sure if it provides equivalent functionality

It does only handle (de)serialization it seems. I do not think using it is worth it, because it basically deserializes into primitive types like `String` and `Vec<Bytes>`.
We should keep it simple and any parsing of JWK parameters is out of scope IMO (or do we want to parse X.509 certificate chains? I dont think so)

**Comment by ubamrein** [comment]
> > There's also https://github.com/RustCrypto/JOSE/tree/master/jose-jwk although offhand I'm not sure if it provides equivalent functionality
> 
> It does only handle (de)serialization it seems. I do not think using it is worth it, because it basically deserializes into primitive types like `String` and `Vec<Bytes>`. We should keep it simple and any parsing of JWK parameters is out of scope IMO (or do we want to parse X.509 certificate chains? I dont think so)

Yes I agree. As I said, really liked the one liner to quickly verify e.g. JWTs :)  

I'll try to come up with a PR, to be sure:

- Use custom derive
- Add `use` as a field (for the sake of best practice)
- Ignore unknown properties 

Does that sound reasonable?

**Comment by Erik1000** [comment]
If you do so, remember the `use` parameter is optional

**Comment by tarcieri** [comment]
I think I'm going to extract the JWK implementation into https://github.com/RustCrypto/JOSE/tree/master/jose-jwk and we can continue to work on it there, which will avoid blocking an `elliptic-curve` release.

I agree it probably doesn't belong in `elliptic-curve`.

Edit: PR is up here: https://github.com/RustCrypto/JOSE/pull/85

**Comment by tarcieri** [comment]
I moved the issue to https://github.com/rustcrypto/jose since that's where the implementation is now located.

The old implementation will be gone as of https://github.com/RustCrypto/traits/pull/1963
